### PR TITLE
Fragment referencing/Query composition support

### PR DESCRIPTION
**Pull request by Poincare**
Adding a way to reference fragments from different components of your application and compose queries using those fragments. See #338 

TODO:

- [x] Update CHANGELOG.md with your change
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

### CODE REVIEW

File: src/QueryManager.ts [hunk 1]
```diff
@@ -202,7 +204,7 @@ export class QueryManager {
     mutation: Document,
     variables?: Object,
     resultBehaviors?: MutationBehavior[],
-  }): Promise<GraphQLResult> {
+  }, fragments: FragmentDefinition[] = []): Promise<GraphQLResult> {
```
Comment (stubailo) [inline src/QueryManager.ts]: Why not just add this to the existing list of options?

File: test/QueryManager.ts [hunk 2]
```diff
@@ -2407,6 +2411,137 @@ describe('QueryManager', () => {
       }, 100);
     });
   });
+
+  describe('fragment referencing', () => {
+    it('should accept a list of fragments and let us reference them through fetchQuery', (done) => {
+      const fragment1 = getFragmentDefinitions(gql`
```
Comment (stubailo) [inline test/QueryManager.ts]: I think we have a `getFragmentDefinition` function that gets a single fragment :]

File: test/client.ts [hunk 3]
```diff
@@ -984,4 +986,370 @@ describe('client', () => {
...
+      const fragmentDoc = gql`
+        fragment authorDetails on Author {
+          firstName
+          lastName
+        }
+        fragment personDetails on Person {
+          name
+        }
+        `;
+      const fragmentDefs = client.fragment(fragmentDoc, []);
```
Comment (stubailo) [inline test/client.ts]: Is there any reason this has an empty array as a second arg? I would think that would be optional.

Comment (Poincare) [inline test/client.ts]: You're right - those empty arrays are residue from when it wasn't optional at one point. I'll get rid of the empty arrays.

File: src/QueryManager.ts (hunk 1 above)
Comment (Poincare) [inline src/QueryManager.ts]: That sounds better - will do.

**Comment by stubailo** [comment]
@Poincare one question I just realized - how will this work with the view integrations? Perhaps it should be a static export of `apollo-client`, since it doesn't really require an instance?

Like:

```js
import { createFragments } from 'apollo-client';

createFragments(gql`
  ...
`);
```

Or something.

@jbaxleyiii do you have any thoughts on how this might or should interact with `react-apollo`? My thought is that maybe we don't need anything React-specific here if it's just a static property you can pass around.

**Comment by Poincare** [comment]
I thought the `client` instance was available to `react-apollo` and `angular-apollo`. If that's the case, we might not need a static export.

Secondly, I think holding the state that describes the fragment names should be specific to a particular instance of `Apollo Client` rather than a global across the package. I'm not sure if there's some way around holding package-global state if we use a static export.

File: src/index.ts [hunk 4]
```diff
@@ -63,6 +68,56 @@ export {
...
+      fragmentDefinitionsMap[fragmentName] = [fragmentDefinition];
+    }
+  });
+
+  return fragments.concat(fragmentDefinitions);
+}
+
+// This function disables the warnings printed about fragment names. One place where this chould be
+// is called when writing unit tests that depend on Apollo Client and use named fragments that may
+// have the Nsame name across different unit tests.
```
Comment (stubailo) [inline src/index.ts]: extra capital N

File: src/index.ts [hunk 5]
```diff
@@ -63,6 +68,56 @@ export {
...
+export function createFragment(doc: Document, fragments: FragmentDefinition[] = []): FragmentDefinition[] {
+  const fragmentDefinitions = getFragmentDefinitions(doc);
+  fragmentDefinitions.forEach((fragmentDefinition) => {
+    const fragmentName = fragmentDefinition.name.value;
+    if (fragmentDefinitionsMap.hasOwnProperty(fragmentName) &&
+        fragmentDefinitionsMap[fragmentName].indexOf(fragmentDefinition) === -1) {
+      // this is a problem because the app developer is trying to register another fragment with
+      // the same name as one previously registered. So, we tell them about it.
+      if (printFragmentWarnings) {
+        console.warn(`Warning: fragment with name ${fragmentDefinition.name.value} already exists.`);
```
Comment (stubailo) [inline src/index.ts]: I think we should expand this warning, and say something about how we enforce that all fragment names are unique. Perhaps we should ask people to read the docs as well.

**Comment by stubailo** [comment]
Looks great other than some small comments

**Comment by Poincare** [comment]
Fixed @stubailo

**Pull request merged**